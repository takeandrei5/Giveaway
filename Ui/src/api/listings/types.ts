import { PaginatedResult } from '../../utils/types';

export type ItemData = { id: string; image: string; title: string; createdAt: Date };

export type CreateListingRequest = {
	title: string;
	description: string;
	category: number;
	images: string[];
};

export type UpdateListingRequest = {
	title: string;
	description: string;
	category: number;
	images: string[];
};

export interface FetchListingsResponse {
	listings: PaginatedResult<ItemData>;
}

export type ListingInformation = {
	id: string;
	title: string;
	description: string;
	category: number;
	createdAt: Date;
	images: string[];
};

export type OwnerInformation = {
	email: string;
	name: string;
	image: string;
};

export type FetchListingDetailsResponse = { listingInfo: ListingInformation } & { ownerInfo: OwnerInformation };
