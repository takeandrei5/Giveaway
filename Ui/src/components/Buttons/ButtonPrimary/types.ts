import { ButtonProps } from '@chakra-ui/react';

export type ButtonPrimaryProps = Omit<ButtonProps, 'color'>;
