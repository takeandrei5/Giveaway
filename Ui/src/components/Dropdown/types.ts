export type DropdownProps = {
	id: string;
	options: DropdownOption[];
	onChangeHandler: (value: string) => void;
	name: string;
	value: string | number | readonly string[];
	isInvalid?: false;
};

export type DropdownOption = {
	value: string | number;
	displayValue: string;
};
