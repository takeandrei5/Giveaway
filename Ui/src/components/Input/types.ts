export type InputProps = {
	id: string;
	name: string;
	placeholder: string;
	value?: string;
	onBlur?: (e: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>) => void;
	onChange?: (e: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>) => void;
	onKeyDown?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	onKeyPress?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	onKeyUp?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	disabled?: boolean;
	multiline?: boolean;
	isInvalid?: boolean;
	height?: string | number;
	label?: string;
	leftIcon?: JSX.Element;
	rightIcon?: JSX.Element;
	rows?: number;
	width?: string | number;
};
