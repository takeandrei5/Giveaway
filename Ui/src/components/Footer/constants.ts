import { StackItem } from './types';

const stackItems: StackItem[] = [
	{
		stackColumnName: 'Product',
		stackColumnItems: ['Overview', 'Features', 'Tutorials', 'Pricing', 'Releases'],
	},
	{
		stackColumnName: 'Company',
		stackColumnItems: ['About', 'Press', 'Careers', 'Contact', 'Partners'],
	},
	{
		stackColumnName: 'Support',
		stackColumnItems: ['Help Center', 'Terms of Service', 'Legal', 'Privacy Policy', 'Status'],
	},
	{
		stackColumnName: 'Follow us',
		stackColumnItems: ['Facebook', 'Twitter', 'Github', 'Instagram', 'Linkedin'],
	},
];

export { stackItems };
