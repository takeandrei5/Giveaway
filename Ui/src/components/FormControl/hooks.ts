import { FieldMetaProps } from 'formik';
import { useEffect, useState } from 'react';

import { useCheckFormIsInvalidProps } from './types';

const useCheckFormIsInvalid = (meta: FieldMetaProps<unknown>): useCheckFormIsInvalidProps => {
	const [isInvalid, setIsInvalid] = useState(false);

	useEffect(() => {
		setIsInvalid(meta.touched && !!meta.error);
	}, [meta.touched, meta.error]);

	return { isInvalid };
};

export { useCheckFormIsInvalid };
