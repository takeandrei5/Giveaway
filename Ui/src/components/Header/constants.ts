const DEFAULT_AVATAR: string = 'https://avatars.dicebear.com/api/male/username.svg';

export { DEFAULT_AVATAR };
