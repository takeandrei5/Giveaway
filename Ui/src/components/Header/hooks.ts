import { NextRouter, useRouter } from 'next/router';
import { useCallback } from 'react';

const useLogin = () => {
	const router: NextRouter = useRouter();

	const handleSignIn: () => Promise<boolean> = useCallback(
		async () => await router.replace('/api/auth/login'),
		[router]
	);
	const handleSignInWithReturnTo: (returnTo: string) => Promise<boolean> = useCallback(
		async (returnTo: string) => await router.replace(`/api/auth/login?returnTo=${returnTo}`),
		[router]
	);
	const handleSignOut: () => Promise<boolean> = useCallback(() => router.replace('/api/auth/logout'), [router]);

	return { handleSignIn, handleSignInWithReturnTo, handleSignOut };
};

export { useLogin };
