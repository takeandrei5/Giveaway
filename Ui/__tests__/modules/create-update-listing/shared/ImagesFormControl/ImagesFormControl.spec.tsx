import { ImagesFormControl } from '@modules/create-update-listing/shared/ImagesFormControl';
import { useImageUpload } from '@modules/create-update-listing/shared/ImagesFormControl/hooks';
import { ImagesFormControlProps } from '@modules/create-update-listing/shared/ImagesFormControl/types';
import { ImageFormikValue } from '@modules/create-update-listing/shared/types';
import { fireEvent, render } from '@testing-library/react';
import { useField } from 'formik';
import { TEST_IMAGE } from '__tests__/constants';
import { QueryClientWrapper } from '__tests__/wrappers';

jest.mock('next/config', () => () => ({
	publicRuntimeConfig: {
		apiUrl: 'http://localhost/',
	},
	serverRuntimeConfig: {
		apiUrl: 'http://localhost/',
	},
}));

jest.mock('formik', () => ({
	...jest.requireActual('formik'),
	useField: jest.fn(),
}));

jest.mock('@modules/create-update-listing/shared/ImagesFormControl/hooks', () => ({
	...jest.requireActual('@modules/create-update-listing/shared/ImagesFormControl/hooks'),
	useImageUpload: jest.fn(),
}));

describe('ImagesFormControl', () => {
	let fieldValue: ImageFormikValue[] = [] as ImageFormikValue[];

	const setTouchedMock: jest.Mock<void, [value: boolean, shouldValidate?: boolean | undefined]> = jest.fn<
		void,
		[value: boolean, shouldValidate?: boolean | undefined]
	>();

	const setValueMock: jest.Mock<void, [value: ImageFormikValue[], shouldValidate?: boolean | undefined]> = jest.fn(
		(value: ImageFormikValue[], shouldValidate?: boolean) => {
			fieldValue = value;
		}
	);

	let onImageUploaded: jest.Mock<void, [id: string, file: File | undefined]> = jest.fn<
		void,
		[id: string, file: File | undefined]
	>();
	let onImageDeleted: jest.Mock<void, [id: string]> = jest.fn<void, [id: string]>();
	let isUploading: boolean = false;

	beforeEach(() => {
		fieldValue = [
			{
				id: 'test-id1',
				url: TEST_IMAGE,
			},
			{
				id: 'test-id2',
				url: TEST_IMAGE,
			},
			{
				id: 'test-id3',
				url: TEST_IMAGE,
			},
		];

		(useField as unknown as jest.Mock).mockImplementation(() => [
			{
				value: fieldValue,
			},
			{
				value: fieldValue,
				error: '',
				touched: false,
				initialValue: fieldValue,
				initialTouched: false,
				initialError: '',
			},
			{
				setValue: setValueMock,
				setTouched: setTouchedMock,
			},
		]);
	});

	it('should match snapshot', () => {
		// Arrange
		const props: ImagesFormControlProps = {
			name: 'test-images',
		};

		(useImageUpload as unknown as jest.Mock).mockImplementation(() => ({
			onImageUploaded,
			onImageDeleted,
			isUploading,
		}));

		// Act
		const { container } = render(<ImagesFormControl {...props} />, { wrapper: QueryClientWrapper });

		// Assert
		expect(container).toMatchSnapshot();
	});

	it('should execute onImageDeleted when the delete image button is clicked', () => {
		// Arrange
		const props: ImagesFormControlProps = {
			name: 'test-images',
		};

		(useImageUpload as unknown as jest.Mock).mockImplementation(() => ({
			onImageUploaded,
			onImageDeleted,
			isUploading,
		}));

		// Act
		const { container } = render(<ImagesFormControl {...props} />, { wrapper: QueryClientWrapper });

		fireEvent.click(container.querySelector('#delete-image-button')!);

		// Assert
		expect(onImageDeleted).toBeCalledTimes(1);
	});

	describe('Spinner', () => {
		it('should render a spinner if isUploading is true', () => {
			// Arrange
			const props: ImagesFormControlProps = {
				name: 'test-images',
			};

			isUploading = true;

			(useImageUpload as unknown as jest.Mock).mockImplementation(() => ({
				onImageUploaded,
				onImageDeleted,
				isUploading,
			}));

			// Act
			const { getByTestId } = render(<ImagesFormControl {...props} />, { wrapper: QueryClientWrapper });

			// Assert
			expect(getByTestId('uploading-image-spinner')).toBeTruthy();
		});

		it('should not render a spinner if isUploading is false', () => {
			// Arrange
			const props: ImagesFormControlProps = {
				name: 'test-images',
			};

			isUploading = false;

			(useImageUpload as unknown as jest.Mock).mockImplementation(() => ({
				onImageUploaded,
				onImageDeleted,
				isUploading,
			}));

			// Act
			const { queryByTestId } = render(<ImagesFormControl {...props} />, { wrapper: QueryClientWrapper });

			// Assert
			expect(queryByTestId('uploading-image-spinner')).toBeFalsy();
		});
	});

	describe('File input', () => {
		it('should have pointer events `none` if isUploading is true', () => {
			// Arrange
			const props: ImagesFormControlProps = {
				name: 'test-images',
			};

			isUploading = true;

			(useImageUpload as unknown as jest.Mock).mockImplementation(() => ({
				onImageUploaded,
				onImageDeleted,
				isUploading,
			}));

			// Act
			const { container } = render(<ImagesFormControl {...props} />, { wrapper: QueryClientWrapper });

			// Assert
			const styles: CSSStyleDeclaration = getComputedStyle(container.querySelector('#file-input')!);

			expect(styles).toHaveProperty('pointer-events', 'none');
		});

		it('should have pointer events `none` if isUploading is false', () => {
			// Arrange
			const props: ImagesFormControlProps = {
				name: 'test-images',
			};

			isUploading = false;

			(useImageUpload as unknown as jest.Mock).mockImplementation(() => ({
				onImageUploaded,
				onImageDeleted,
				isUploading,
			}));

			// Act
			const { container } = render(<ImagesFormControl {...props} />, { wrapper: QueryClientWrapper });

			// Assert
			const styles: CSSStyleDeclaration = getComputedStyle(container.querySelector('#file-input')!);

			expect(styles).toHaveProperty('pointer-events', 'auto');
		});
	});
});
