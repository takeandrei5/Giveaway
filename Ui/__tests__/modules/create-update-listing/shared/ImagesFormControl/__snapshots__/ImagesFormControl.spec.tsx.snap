// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImagesFormControl should match snapshot 1`] = `
<div>
  <div
    class="chakra-form-control css-8atqhb"
    role="group"
  >
    <label
      class="chakra-form__label css-xgxc5v"
      for="images-dragdrop"
      id="field-1-label"
    >
      Images:
      <span
        aria-hidden="true"
        class="chakra-form__required-indicator css-0"
        role="presentation"
      >
        *
      </span>
    </label>
    <div
      class="css-11bgua5"
      data-rbd-droppable-context-id="0"
      data-rbd-droppable-id="droppable"
    >
      <div
        aria-describedby="rbd-hidden-text-0-hidden-text-0"
        class="css-1de6poj"
        data-rbd-drag-handle-context-id="0"
        data-rbd-drag-handle-draggable-id="test-id1"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="test-id1"
        draggable="false"
        role="button"
        tabindex="0"
      >
        <div
          class="css-lkih57"
        >
          <button
            class="chakra-button css-arwpy2"
            id="delete-image-button"
            type="button"
          >
            <svg
              class="chakra-icon css-1gr4eat"
              fill="currentColor"
              focusable="false"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4,5 L20,5 L20,23 L4,23 L4,5 Z M1,5 L23,5 M9,1 L15,1 L15,5 L9,5 L9,1 Z M9,1 L15,1 L15,5 L9,5 L9,1 Z M15,9 L15,19 M9,9 L9,19"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
        <input
          accept="image/png, image/jpeg"
          class="chakra-input css-1jccxn"
          id="file-input"
          style="text-indent: -999px;"
          title="Upload an image"
          type="file"
        />
        <svg
          class="chakra-icon css-7l0q8k"
          fill="currentColor"
          focusable="false"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M2 15.5V2a2 2 0 0 1 2-2h8a2 2 0 0 1 2 2v13.5a.5.5 0 0 1-.74.439L8 13.069l-5.26 2.87A.5.5 0 0 1 2 15.5zM8.16 4.1a.178.178 0 0 0-.32 0l-.634 1.285a.178.178 0 0 1-.134.098l-1.42.206a.178.178 0 0 0-.098.303L6.58 6.993c.042.041.061.1.051.158L6.39 8.565a.178.178 0 0 0 .258.187l1.27-.668a.178.178 0 0 1 .165 0l1.27.668a.178.178 0 0 0 .257-.187L9.368 7.15a.178.178 0 0 1 .05-.158l1.028-1.001a.178.178 0 0 0-.098-.303l-1.42-.206a.178.178 0 0 1-.134-.098L8.16 4.1z"
            fill-rule="evenodd"
          />
        </svg>
        <div
          class="css-s0voyp"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
          >
            <span
              style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 100% 0px 0px 0px;"
            />
            <img
              alt="imageUpload"
              data-nimg="responsive"
              decoding="async"
              draggable="false"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
            />
            <noscript />
          </span>
        </div>
      </div>
      <div
        aria-describedby="rbd-hidden-text-0-hidden-text-0"
        class="css-1de6poj"
        data-rbd-drag-handle-context-id="0"
        data-rbd-drag-handle-draggable-id="test-id2"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="test-id2"
        draggable="false"
        role="button"
        tabindex="0"
      >
        <div
          class="css-lkih57"
        >
          <button
            class="chakra-button css-arwpy2"
            id="delete-image-button"
            type="button"
          >
            <svg
              class="chakra-icon css-1gr4eat"
              fill="currentColor"
              focusable="false"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4,5 L20,5 L20,23 L4,23 L4,5 Z M1,5 L23,5 M9,1 L15,1 L15,5 L9,5 L9,1 Z M9,1 L15,1 L15,5 L9,5 L9,1 Z M15,9 L15,19 M9,9 L9,19"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
        <input
          accept="image/png, image/jpeg"
          class="chakra-input css-1jccxn"
          id="file-input"
          style="text-indent: -999px;"
          title="Upload an image"
          type="file"
        />
        <div
          class="css-s0voyp"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
          >
            <span
              style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 100% 0px 0px 0px;"
            />
            <img
              alt="imageUpload"
              data-nimg="responsive"
              decoding="async"
              draggable="false"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
            />
            <noscript />
          </span>
        </div>
      </div>
      <div
        aria-describedby="rbd-hidden-text-0-hidden-text-0"
        class="css-1de6poj"
        data-rbd-drag-handle-context-id="0"
        data-rbd-drag-handle-draggable-id="test-id3"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="test-id3"
        draggable="false"
        role="button"
        tabindex="0"
      >
        <div
          class="css-lkih57"
        >
          <button
            class="chakra-button css-arwpy2"
            id="delete-image-button"
            type="button"
          >
            <svg
              class="chakra-icon css-1gr4eat"
              fill="currentColor"
              focusable="false"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4,5 L20,5 L20,23 L4,23 L4,5 Z M1,5 L23,5 M9,1 L15,1 L15,5 L9,5 L9,1 Z M9,1 L15,1 L15,5 L9,5 L9,1 Z M15,9 L15,19 M9,9 L9,19"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
        <input
          accept="image/png, image/jpeg"
          class="chakra-input css-1jccxn"
          id="file-input"
          style="text-indent: -999px;"
          title="Upload an image"
          type="file"
        />
        <div
          class="css-s0voyp"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
          >
            <span
              style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 100% 0px 0px 0px;"
            />
            <img
              alt="imageUpload"
              data-nimg="responsive"
              decoding="async"
              draggable="false"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
            />
            <noscript />
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
